service/dap/server.go

```diff
@@ -415,6 +424,45 @@ func (s *Server) handleRequest(request dap.Message) {
 				Body:     dap.ThreadsResponseBody{Threads: []dap.Thread{{Id: -1, Name: "Current"}}},
 			}
 			s.send(response)
+		case *dap.SetBreakpointsRequest:
+			// Halt running command
+			inProgress := s.debugger.RunningCommand()
+			// The channel is drained before running a command, but just in case drain again to avoid blocking
+			s.resetOverrideStopReason()
+			// TODO(polina): once api.Call is run asynchronously, treat it like api.Continue as well
+			if inProgress == api.Continue {
+				s.overrideStopReason <- skipStop
```

Comment:
I don't think this is a good way of doing this. By the time you send the halt command the target might be already in the process of stopping for a different reason, so you can't blindly skip the next stop, because it could be one that the user wants to see. You have to issue the halt command, examine the state that is returned to you to determine if you are the cause of the stop or not and then either:
- suppress the stop event
- set breakpoints
- automatically resume execution
or:
- propagate the stop event
- set breakpoints

Doing this correctly is complicated because you have to synchronize with both your own goroutines (which you have to do with mutexes and the likes) as well as with target threads (which you have to do by post-facto examination of its state). Personally, I wouldn't get into it, too complicated for such a marginally useful feature.

Reply:
The approach in this PR was inspired by how the original adapter handled this feature, which was clearly not quite right :)

I am hesitant not to support this at all because vscode users are used to this working.

Instead of asking the users to manually halt to set breakpoints, a compromise solution could be to support halt+setBreakpoints only and let the user continue as they see fit. This will save the users a step, and make it obvious what is going on.

As for understanding why the command stopped, this is something we might want to do a better job at anyway since we need to communicate the stop reasons and the details to the user and not just here, but in case of any stop. This could be done in a separate PR, but let me clarify the details here while we are on topic. Plus I could then add a TODO in this PR for what it would take to resume execution after setting a breakpoint while running.

If debugger.stopReason() is "manual", should we still be going over the threads and looking for any with a breakpoint? If the stop reason is anything other than "next finished", should we be checking NextInprogress==true and reporting (e.g. as a message in the Console) that a next operation is in progress and that hitting continue would complete it?

Reply:
>If debugger.stopReason() is "manual", should we still be going over the threads and looking for any with a breakpoint? If the stop reason is anything other than "next finished", should we be checking NextInprogress==true and reporting (e.g. as a message in the Console) that a next operation is in progress and that hitting continue would complete it?

No. You should stop (i.e. send a stop event and not resume the process):

1. If StopReason is anything but manual
2. If any thread has a breakpoint or CallReturn set
3. If NextInProgress is false *and* the last command sent *by the user* is: next, step, stepOut, reverseNext, reverseStep or reverseStepOut

If you *do not* stop you should suppress the stop event that continue would send and execute the api.DirectionCongruentContinue command. This presumes that the "halt-without-cancelling-next" is implemented and used.
Letting the stop event go through and sending a new resume event doesn't work without explicit synchronization (the client could get them in the wrong order).

Reply:
Thank you, this is very helpful.
I have added a TODO for resuming execution with this information. The scope of this CL is resumed to just halt+set.
I will check what can be done to improve stopped event reporting in a different PR.